### PR TITLE
Fix immutable sluggable behavior

### PR DESCRIPTION
**Pull request by Kolyunya**
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #13728

`isNewSlugNeeded()` wasn't actually checked when `attribute` was equal to `null`.

### CODE REVIEW

**Review by samdark — APPROVED** [review]

**Review by SilverFire — APPROVED** [review]

File: framework/behaviors/SluggableBehavior.php [hunk 1]
```diff
@@ -171,6 +171,10 @@ protected function isNewSlugNeeded()
             return false;
         }
 
+        if ($this->attribute === null) {
```
Comment (dynasource) [inline framework/behaviors/SluggableBehavior.php]: this feels a bit magical and is taking too much time to understand. Can you support this with a comment to explain why true is to be returned when attribute is null?

**Review by dynasource — CHANGES_REQUESTED** [review]

File: framework/behaviors/SluggableBehavior.php (hunk 1 above)
Comment (Kolyunya) [inline framework/behaviors/SluggableBehavior.php]: Wouldn't [this](https://github.com/yiisoft/yii2/pull/13747) be sufficient?

Comment (Kolyunya) [inline framework/behaviors/SluggableBehavior.php]: Sorry I missed the line.

Comment (Kolyunya) [inline framework/behaviors/SluggableBehavior.php]: Seems like we can optimize this condition for the case when `$value` is a string. We can compare it to the current slug value. Do you think it worth implementing?

For the case when `$value` is `callable` the comment would be:
> Callback-based non-immutable slugs should be recalculated each time.

Comment (dynasource) [inline framework/behaviors/SluggableBehavior.php]: well, the current state of SluggableBehavior is too vague IMO, so every step to make it more readable is a step in the right direction.

Comment (Kolyunya) [inline framework/behaviors/SluggableBehavior.php]: Should we add another condition for the case when `$value` is a string or just recalculate all value-based slugs?

Comment (dynasource) [inline framework/behaviors/SluggableBehavior.php]: what would be the usecase for that?

Comment (Kolyunya) [inline framework/behaviors/SluggableBehavior.php]: When the slug is string-based and it's value is equal to the `$value` we can return `false` and slug value won't be recalculated.

Comment (dynasource) [inline framework/behaviors/SluggableBehavior.php]: if this condition had to be covered by a method with a proper self declaring name, how would you call this method?

Comment (Kolyunya) [inline framework/behaviors/SluggableBehavior.php]: Do you think we should move this condition to a separate method?

I added a WIP comment. The problem is that it tells that `callback-based` slugs are always reevaluated but in fact all `value-based` slugs are always reevaluated at the moment.

Comment (dynasource) [inline framework/behaviors/SluggableBehavior.php]: Im trying to get concrete answers to my questions to get a proper understanding ;)

Comment (Kolyunya) [inline framework/behaviors/SluggableBehavior.php]: Could you please ask a *self-contained* concrete question please? What particularly don't you understand?

Comment (dynasource) [inline framework/behaviors/SluggableBehavior.php]: We want to know 'what' the software does and not 'how' he does it. 

I tried to get to this, but I could not get clear answers to my questions. Examples:

* My question "Can you support this with a comment to explain why true is to be returned when attribute is null" in https://github.com/yiisoft/yii2/pull/13729#discussion_r106017513 gets a vague implementation reaction in https://github.com/yiisoft/yii2/pull/13729#discussion_r106027562
* My question "what would be the usecase for that" gets an answer about implemantation details in https://github.com/yiisoft/yii2/pull/13729#discussion_r106033797 but not about the "why" which was my question
* My last question to understand the "why" https://github.com/yiisoft/yii2/pull/13729#discussion_r106035189 was about trying to get you to formulate the usecase into a understandable testMethod, but again I dont get an corresponding answer to my question in https://github.com/yiisoft/yii2/pull/13729#discussion_r106101040

To go back to the code. My problem is to understand the logic behind `$this->attribute === null`. What is the meaning of it. Why is it added? By purely looking at the code, I have to put effort into understand it. This goes against the best practice that code should be self-explanatory and instantly understandable. 

So again. We want to know 'what' the software does and not 'how' he does it.

Comment (Kolyunya) [inline framework/behaviors/SluggableBehavior.php]: > Can you support this with a comment to explain why true is to be returned when attribute is null?
Yes I can. And I did it here: https://github.com/yiisoft/yii2/pull/13729/files#r106027562

> What would be the usecase for that?
String-based slugs.

> What is the meaning of it. Why is it added?
It seems to be obvious for me. It's not possible to tell if new slug is needed or not for callback-based slugs. That is why a new value is always needed for such slugs.

**Comment by samdark** [comment]
Merged. Thank you!

**Comment by Kolyunya** [comment]
Glad to help!

**Pull request merged**